### ISSUE TITLE
[Question] How to align the predicted depths to a real depth map.

### ISSUE THREAD

**Issue by Master-cai** [DepthAnything/Depth-Anything-V2#157]
Hi guys, I want to improve a low-resolution depth map by aligning it to the output of Depth-anything-v2, How can I make it? Should I use the output of the default models(relative depth estimation) or the metric_depth? the predicted depth map is the scaled and shifted version of real depth image, right? 

thanks!

cc @LiheYoung @bingykang @1ssb 

[related issue](https://github.com/isl-org/MiDaS/issues/171)

**Comment by LiheYoung** [comment]
Hi @Master-cai, you can try both alternatives by fitting the scale and shift values to map the model's predictions to real metric depth. If you're using our relative depth predictions, please remember to invert the values (originally in the disparity space) back to the depth space first.

**Comment by Master-cai** [comment]
@LiheYoung Hi, I use this equation: `(1. / metric_depth) = shift + (scale * output)` to compute the scale and shift via `lstsq`, and get the metric by `1. / (shift + (scale * output))`. The `output` means the output of the `vitl` model. 

The results seems good on several test data. 

I want to confirm my equation is correct and it can be a reference for whoever want to do the samilar things.

**Comment by LiheYoung** [comment]
Hi @Master-cai, your equation is correct.

**Comment by Vincent630** [comment]
> @LiheYoung Hi, I use this equation: `(1. / metric_depth) = shift + (scale * output)` to compute the scale and shift via `lstsq`, and get the metric by `1. / (shift + (scale * output))`. The `output` means the output of the `vitl` model.
> 
> The results seems good on several test data.
> 
> I want to confirm my equation is correct and it can be a reference for whoever want to do the samilar things.

hello,i want to know how you get the value of scale and shift?and when you do as you said, how is it going and how much the deviation between caculated metric depth and real depth map？maybe 5-10cm?
thanks a lot, please let me know if that is ok for you

**Comment by Master-cai** [comment]
@Vincent630 Hi! I  get the scale and shift by least squares method (lstsq), as above mentioned. 
>how is it going and how much the deviation between caculated metric depth and real depth map？maybe 5-10cm?

I do not align to real depth map so I can't answer this question, but in my experiment, the accuracy is enough.

Hope it's helpful

**Comment by I3aer** [comment]
I think the correction equation is 1/Depth = 1/(a*Relative + b) if you are using the relative models.

**Comment by ForestWang** [comment]
> [@LiheYoung](https://github.com/LiheYoung) Hi, I use this equation: `(1. / metric_depth) = shift + (scale * output)` to compute the scale and shift via `lstsq`, and get the metric by `1. / (shift + (scale * output))`. The `output` means the output of the `vitl` model.
> 
> The results seems good on several test data.
> 
> I want to confirm my equation is correct and it can be a reference for whoever want to do the samilar things.

could you explain how to calculate the shift and scale parameters. from the equation `(1. / metric_depth) = shift + (scale * output)` , **output** is the model's output result, and we dont know the **metric_depth** value.

**Comment by Master-cai** [comment]
> > [@LiheYoung](https://github.com/LiheYoung) Hi, I use this equation: `(1. / metric_depth) = shift + (scale * output)` to compute the scale and shift via `lstsq`, and get the metric by `1. / (shift + (scale * output))`. The `output` means the output of the `vitl` model.
> > The results seems good on several test data.
> > I want to confirm my equation is correct and it can be a reference for whoever want to do the samilar things.
> 
> could you explain how to calculate the shift and scale parameters. from the equation `(1. / metric_depth) = shift + (scale * output)` , **output** is the model's output result, and we dont know the **metric_depth** value.

In my case, the "metric_depth" is given.

**Comment by luisdecker** [comment]
@Master-cai Does the "lstsq" method you mention is the closed least squares equation proposed on MIDAS? In the case, 

```python
def compute_scale_and_shift(self, prediction, target, mask):
        # system matrix: A = [[a_00, a_01], [a_10, a_11]]
        a_00 = torch.sum(mask * prediction * prediction, (1, 2))
        a_01 = torch.sum(mask * prediction, (1, 2))
        a_11 = torch.sum(mask, (1, 2))

        # right hand side: b = [b_0, b_1]
        b_0 = torch.sum(mask * prediction * target, (1, 2))
        b_1 = torch.sum(mask * target, (1, 2))

        # solution: x = A^-1 . b = [[a_11, -a_01], [-a_10, a_00]] / (a_00 * a_11 - a_01 * a_10) . b
        x_0 = torch.zeros_like(b_0)
        x_1 = torch.zeros_like(b_1)

        det = a_00 * a_11 - a_01 * a_01
        # A needs to be a positive definite matrix.
        valid = det > 0

        x_0[valid] = (a_11[valid] * b_0[valid] - a_01[valid] * b_1[valid]) / det[valid]
        x_1[valid] = (-a_01[valid] * b_0[valid] + a_00[valid] * b_1[valid]) / det[valid]

        return x_0, x_1
```
??
If not, could you please provide a gist with your code?

**Comment by Master-cai** [comment]
> If not, could you please provide a gist with your code?

Hope the code snippet help.

```python
    Y = Y.view(-1, 1)
    X = X.view(-1, 1)
    ones = torch.ones_like(X)
    A_B = torch.cat([ones, Y], dim=1)

    A_B = A_B.cpu()
    X = X.cpu()
    solution = torch.linalg.lstsq(A_B, 1.0 / X).solution
    solution = solution.to(trg_depth.device)
    A, B = solution[:2].squeeze()

    output = 1.0 / (A + B * ref_depth)
    
    return output, A, B
```

**Comment by ihan1208** [comment]
> > If not, could you please provide a gist with your code?
> 
> Hope the code snippet help.
> 
>     Y = Y.view(-1, 1)
>     X = X.view(-1, 1)
>     ones = torch.ones_like(X)
>     A_B = torch.cat([ones, Y], dim=1)
> 
>     A_B = A_B.cpu()
>     X = X.cpu()
>     solution = torch.linalg.lstsq(A_B, 1.0 / X).solution
>     solution = solution.to(trg_depth.device)
>     A, B = solution[:2].squeeze()
> 
>     output = 1.0 / (A + B * ref_depth)
>     
>     return output, A, B

If we do not have the ground truth depth, how can we build this lstsp （I.E., we only have the 'output', but, without the 'depth_metric'）?

**Comment by luisdecker** [comment]
> > > If not, could you please provide a gist with your code?
> > 
> > 
> > Hope the code snippet help.
> > ```
> > Y = Y.view(-1, 1)
> > X = X.view(-1, 1)
> > ones = torch.ones_like(X)
> > A_B = torch.cat([ones, Y], dim=1)
> > 
> > A_B = A_B.cpu()
> > X = X.cpu()
> > solution = torch.linalg.lstsq(A_B, 1.0 / X).solution
> > solution = solution.to(trg_depth.device)
> > A, B = solution[:2].squeeze()
> > 
> > output = 1.0 / (A + B * ref_depth)
> > 
> > return output, A, B
> > ```
> 
> If we do not have the ground truth depth, how can we build this lstsp （I.E., we only have the 'output', but, without the 'depth_metric'）?

If you don't have a gt depth, you can only have a relative depth estimation. This alignment is useful just to compute metrics or to compare with the GT.

**Comment by RoyAmoyal** [comment]
> > > > If not, could you please provide a gist with your code?
> > > 
> > > 
> > > Hope the code snippet help.
> > > ```
> > > Y = Y.view(-1, 1)
> > > X = X.view(-1, 1)
> > > ones = torch.ones_like(X)
> > > A_B = torch.cat([ones, Y], dim=1)
> > > 
> > > A_B = A_B.cpu()
> > > X = X.cpu()
> > > solution = torch.linalg.lstsq(A_B, 1.0 / X).solution
> > > solution = solution.to(trg_depth.device)
> > > A, B = solution[:2].squeeze()
> > > 
> > > output = 1.0 / (A + B * ref_depth)
> > > 
> > > return output, A, B
> > > ```
> > 
> > 
> > If we do not have the ground truth depth, how can we build this lstsp （I.E., we only have the 'output', but, without the 'depth_metric'）?
> 
> If you don't have a gt depth, you can only have a relative depth estimation. This alignment is useful just to compute metrics or to compare with the GT.

Is it useful to use it for a single image with ground truth of a scene, and then use the scale and shift for the rest of the images of the same scene? Did you evaluate this approach? Thanks!
